### PR TITLE
Add file offset

### PR DESCRIPTION
**Pull request by joshuak94**
Adds a new functionality to `sam_file_input`, namely the ability to store the file offset of a read, and then later to seek to the given file offset. This is done by adding `seqan3::field::file_offset` as a field, and adding the function `sam_file_input::seek`.

### CODE REVIEW

**Comment by marehr** [comment]
@joshuak94 I put this decision in our next strategy meeting

@seqan/core  I put that into triage board, please remember me that we have to discuss this :)

**Comment by joshuak94** [comment]
@marehr Sounds good!

**Comment by marehr** [comment]
> By the way, I've done some more reading on `std::streampos` and I was wondering if you could explain `std::streampos` vs `std::streamoff`. I see that the former stores also the "shift state", but I'm not really sure what that means and if it's really necessary for our case, because then I think I'd rather use `std::streamoff` here.

The thing is that you need both (offset + state) to make sure to get the same absolute position, and you are on the other hand right, that we always use a mode that should be agnostic to this (mostly, because we always try to read byte's).

I can understand that it might be a waste of space if you store many of them (I tried it out and `sizeof(std::streampos) == sizeof(std::streamoff) + 8` which doubles the size on my system), but you are always able to convert it into a `std::streamoff`.

I'm in slight favour for `std::streampos` as return type, as this has the specific semantic of an absolute position into a file (which might be different from the number of bytes read). Furthermore, I'm not too sure if relative offset-positions will always be correctly implemented as functions like [seekg](https://en.cppreference.com/w/cpp/io/basic_istream/seekg) primarily expect absolute positions (our bgzf-stream for example only supports stream offsets from the beginning of a file, other io implementations might only support some certain offset-directions).

> Interestingly, BAM index format stores the offsets int `uint64_t`, even though -1 is a valid value for it.

Can you explain this a bit more? Why is `-1` a valid value? I mean `-1` will underflow and produces `2^64-1`, of course it is a valid value, but is it an unexpected value or is it a special value?

**Comment by joshuak94** [comment]
I mean that `std::streampos` and `std::streamoff` both can have a value of -1, according to [cppreference](https://en.cppreference.com/w/cpp/io/fpos). But in the SAM format standard, they store their file offsets as uint64_t types.

**Comment by marehr** [comment]
### Core-Meeting 2021-04-12

Hi all, we have the following idea to make the change smaller in scope, but as usable as you want it to be:

* We want to use `std::streampos` instead of `std::streamoff`, as some implementations, like BGZF streams, have no meaningful "offset" implementation. Thus, we want the naming `file_position` instead of `file_offset`, as it is an absolute position into a file and not some offset between two positions.
* We want to have that seek / position feature as a low-level API, that sits directly in the iterator.
  (If we ever want to support memory mapping, setting a global position in the file itself seems like some unnecessary global state to store)
* We found the hidden requirement, that all "secondary streams" (e.g. BGZF uncompressing) must support global positioning.

I hope this example of the usage is enough to show how the interface should be defined:

Before:
```c++
sam_file.seek(position);
for (auto && record: sam_file)
{
  record.file_offset();
}
```

Now:
```c++
auto it = sam_file.begin(); // reads in header, is now at the position of the first record
auto end = sam_file.end();
for (; it != end; ++it)
{
    // the follwing should have no "effect"
    // (of course it re-seeks internally and is a waste of time, but the returned record is the same :))
    it.seek_to(it.file_position());
    *it; // returns record
     
    *it.seek_to(it.file_position()); // returns the record, i.e. seek_to() -> iterator&; returns the iterator itself

    it.seek_to(some_other_position); // the iterator will be seeked to that position and ++it will go on from there
}
```

That means the iterator gets two new functions
* We think you can add that functionality to `include/seqan3/io/detail/in_file_iterator.hpp`
* As that iterator is currently non-public API, we keep that functionality as a "hidden" feature.
* `it.file_position() -> std::streampos` give current file (absolute) position of the record 
  * document the function as a low-level API
* `it.seek_to(std::streampos{50}) -> iterator &`  seek to the given position and read / buffer the record
  * note: header must not be parsed, as sam_file.begin() already did it
  * after seeking, check the failbit https://en.cppreference.com/w/cpp/io/basic_ios/fail; if set throw an exception 
  * document the function as a low-level API
* Any suggestions? Have we missed something?

In any case if you have any question we can talk about the changes :)

**Comment by joshuak94** [comment]
@marehr One question! For the tests, where should I put them? The in_file_iterator_test only has a dummy type, which I'm pretty sure will not work with the `tellg()` function... I could test it in sam_file_input_test? Should I then also test it in the sequence_file and structure_file test directories?

Actually, is there any way within the `file_position()` function to make sure `host` is an ifstream?

**Comment by marehr** [comment]
> @marehr One question! For the tests, where should I put them? The in_file_iterator_test only has a dummy type, which I'm pretty sure will not work with the `tellg()` function... 

You are right, it seems that we need to adapt that test to actually handle a seqan3 "input_file". It seems that iterator implementation requires the following things now:

* file_type has a member function `host->read_next_record()`
* file_type has a member `host->record_buffer`
* file_type has a member `host->at_end`
* file_type has a member `host->secondary_stream` which is a `std::istream` (NEW requirement)

I looked at it, and you would need to change the underlying data structure from `std::vector<int>` to `std::istream` and a bit how the fake_file is being constructed. I can help you if you want :)


> In sam_file_input_test it seems the tests are all done with an `istringstream` as input, not a file (outside of empty files). Do we have any tests which parse actual files?

That seems to be missing right now in our test infrastructure. I would suggest adding a new test case (I would create a new file) that specifically tests this on a manually crafted BAM file that has maybe 3 alignments stored, where each alignment is in its own BGZF-Block. I.e. a BAM File with one BGZF-Block for the header, three BGZF-Blocks for three alignments and one STOP-BGZF-Block. I can help you with that if you need help :)

(That will show that the expected positions are not byte-positions, but a special addressing scheme)

> 
> Actually, is there any way within the `file_position()` function to make sure `host` is an ifstream?

**Comment by joshuak94** [comment]
> > @marehr One question! For the tests, where should I put them? The in_file_iterator_test only has a dummy type, which I'm pretty sure will not work with the `tellg()` function...
> 
> You are right, it seems that we need to adapt that test to actually handle a seqan3 "input_file". It seems that iterator implementation requires the following things now:
> 
>     * file_type has a member function `host->read_next_record()`
> 
>     * file_type has a member `host->record_buffer`
> 
>     * file_type has a member `host->at_end`
> 
>     * file_type has a member `host->secondary_stream` which is a `std::istream` (NEW requirement)
> 
> 

Right, is there a place for me to specifically require these things? From what I see of the iterator class, `file_type` doesn't seem to have any explicit requirements.

> I looked at it, and you would need to change the underlying data structure from `std::vector<int>` to `std::istream` and a bit how the fake_file is being constructed. I can help you if you want :)
> 
> > In sam_file_input_test it seems the tests are all done with an `istringstream` as input, not a file (outside of empty files). Do we have any tests which parse actual files?
> 
> That seems to be missing right now in our test infrastructure. I would suggest adding a new test case (I would create a new file) that specifically tests this on a manually crafted BAM file that has maybe 3 alignments stored, where each alignment is in its own BGZF-Block. I.e. a BAM File with one BGZF-Block for the header, three BGZF-Blocks for three alignments and one STOP-BGZF-Block. I can help you with that if you need help :)
> 
> (That will show that the expected positions are not byte-positions, but a special addressing scheme)

I might need help with creating this file...!

**Comment by joshuak94** [comment]
@marehr I'm probably doing this incorrectly, but here's what my current implementation looks like:

```cpp
     //!\brief Returns the current position in the file via `std::streampos`.
     std::streampos const file_position()
     {
         assert(host != nullptr);
         return host->tellg();
     }

     //!\brief Low level API. Sets the current position of the iterator to the given position, throws if at end of file.
     in_file_iterator & seek_to(std::streampos const & pos)
     {
         assert(host != nullptr);
         host->seekg(pos);
         host->read_next_record();
         if (host->fail())
         {
             throw std::exception("Seeking to file position failed!\n");
         }
         return *this;
     }
```

I"m not sure why host needs to have a secondary stream. To be honest, I'm not quite sure how the secondary_stream/primary_stream are interacting with one another. Are you saying that in my above code, I should have 
```cpp
        // In file_position()
        return host->secondary_stream->tellg();
        
        // In seek_to()
        host->secondary_stream->seekg(pos);
```
Like this?

**Comment by marehr** [comment]
> Right, is there a place for me to specifically require these things? From what I see of the iterator class, `file_type` doesn't seem to have any explicit requirements.

You are right, `seqan3::detail::in_file_iterator` is a non-public API, and `file_type` has this hard requirement only implicitly. The description of that class is also completely out-dated. xD

> I might need help with creating this file...!

Can you give me a sam_file that you want to test? I can convert it into such a bam file. We could also re-use a sam_file test. Maybe this one https://github.com/seqan/seqan3/blob/a3a426739ad07531e54fc076dbaea95f6b70578a/test/unit/io/sam_file/format_sam_test.cpp#L34-L40 ?

> @marehr I'm probably doing this incorrectly, but here's what my current implementation looks like:
> I"m not sure why host needs to have a secondary stream. To be honest, I'm not quite sure how the secondary_stream/primary_stream are interacting with one another. Are you saying that in my above code, I should have
> 
> ```c++
>         // In file_position()
>         return host->secondary_stream->tellg();
>         
>         // In seek_to()
>         host->secondary_stream->seekg(pos);
> ```
> 
> Like this?

This looks right :)

**Comment by marehr** [comment]
@joshuak94 Here are the files

[simple_three_verbose_reads.bam.gz](https://github.com/seqan/seqan3/files/6307085/simple_three_verbose_reads.bam.gz)
[simple_three_verbose_reads.bam.note.gz](https://github.com/seqan/seqan3/files/6307086/simple_three_verbose_reads.bam.note.gz)
[simple_three_verbose_reads.sam.gz](https://github.com/seqan/seqan3/files/6307087/simple_three_verbose_reads.sam.gz)

**Comment by joshuak94** [comment]
@marehr  Here is the first implementation of this... I'm 100% sure the new test file is not the best way to do that.

**Comment by joshuak94** [comment]
@marehr I'm not sure what the best way is to change the tests such that they all pass... You probably have a better idea!

**Comment by joshuak94** [comment]
@marehr I think it could be convenient to have an easy way to "reset" an input file/iterator such that it's pointing to the first read again. Thoughts on this? I could implement a `seek_to_beginning` or something similar for the iterator. This would require storing the streampos of the first record within the input object.

**Comment by marehr** [comment]
> @marehr I rebased it... Still having the issues with the CI tests though, if you have any solutions!

I think I found it, you changed the ranges-v3 submodule (I guess because you forgot to use `git submodule update` after the new checkout and committed that accidentally) that introduced that unrelated error you got within the CI.  

> @marehr I think it could be convenient to have an easy way to "reset" an input file/iterator such that it's pointing to the first read again. Thoughts on this? I could implement a `seek_to_beginning` or something similar for the iterator. This would require storing the streampos of the first record within the input object.

Just a stupid thought, but if you want this, you can simply store the position of the first record, right?

**Comment by joshuak94** [comment]
@marehr Thanks so much for your help! Just wondering regarding the custom SAM hash function, why was it necessary? And was it related to the file position stuff?

**Comment by marehr** [comment]
> @marehr Thanks so much for your help! Just wondering regarding the custom SAM hash function, why was it necessary? And was it related to the file position stuff?

It fixes a different problem and could be its own PR. Maybe I should split it from this PR.

**Review by Irallia — CHANGES_REQUESTED** [review]
Two style things maybe, the rest looks good to me!

**Comment by marehr** [comment]
@Irallia Sorry you were too fast, and I rebased in the meantime to clean up the history. I hope this is fine. I added a new seek test for the sequences.

**Review by Irallia — APPROVED** [review]
LGFM

**Comment by marehr** [comment]
@SGSSGene Can you do 2nd review?

**Review by SGSSGene — CHANGES_REQUESTED** [review]
@marehr does a change like this require a changelog entry?

File: include/seqan3/io/detail/in_file_iterator.hpp [hunk 1]
```diff
@@ -152,6 +152,33 @@ class in_file_iterator
     }
     //!\}
 
+    /*!\name File position functionality
+     * \brief Low level API. Enables seeking to and returning specific file positions from the iterator.
+     * \{
+     */
+
+     //!\brief Returns the current position in the file via `std::streampos`.
+     std::streampos const file_position()
```
[the remark this answers is not in the document]
Comment (marehr) [reply include/seqan3/io/detail/in_file_iterator.hpp:161]: Good catch :)

**Comment by marehr** [comment]
> @marehr does a change like this require a changelog entry?

No, currently it is "noapi" / detail code and is only a "feature" for us.

**Review by SGSSGene — APPROVED** [review]
Looks good! (sorry for delaying this PR so much)

**Pull request merged**